### ISSUE TITLE
Mistaken PossiblyNullArgument

### ISSUE THREAD

**Issue by ssnepenthe** [vimeo/psalm#189]
This might speak to my coding style more than anything, but bear with me a moment...

Given the following:

```php
/**
 * @param  mixed  $values
 * @return boolean
 */
function is_array_of_strings($values) : bool
{
    return is_array($values) && ! count(array_filter(
        $values,
        /**
         * @param  mixed $value
         * @return boolean
         */
        function ($value) : bool {
            return ! is_string($value);
        }
    ));
}

/**
 * @param  array|null $values
 * @return array
 */
function act_on_possibly_null_value(?array $values) : array
{
    if (! is_array_of_strings($values)) {
        return [];
    }

    return array_map('trim', $values);
}
```

Psalm is reporting:

```
ERROR: PossiblyNullArgument - test.php:31 - Argument 2 of array_map cannot be null, possibly null value provided
	return array_map('trim', $values);
```

But at the point `array_map` is called, it should be guaranteed that we have an array.

It seems to work fine if I lose the `is_array_of_strings` function and inline the conditions.

**Comment by muglug** [comment]
A similar issue was raised here: https://github.com/phpstan/phpstan/issues/201

The problem is that it's basically impossible to do the necessary analysis in a reasonable amount of time, if every time you hit a function call you branch off to determine its actions on the input data (the example you give is relatively trivial, but it's easy to see how this behaviour could be more complex).

Workarounds:
- Adding `/** @var string[] $values */` above the last `return` call
- inlining the method's contents

----

If this was a very common pattern I might be minded to add some sort of
```
@psalm-assert string[] $values
```
annotation, but that would only really hold if that method threw an exception (vs returning a boolean)

**Comment by ssnepenthe** [comment]
Fair enough. I'll just close this issue then.

> Adding /** @var string[] $values */ above the last return call

Thanks for that - I didn't realize it was possible to override type in this way. I guess I should have read the wiki a little more carefully...
